app/services/share_progress_variant_builder.rb

```diff
@@ -1,17 +1,17 @@
 require 'share_progress'
 
 class ShareProgressVariantBuilder
-  def self.create(params, variant_type:, page:, url:)
-    new(params, variant_type, page, url).create
+  def self.create(params:, variant_type:, page:, url:)
+    new(params, variant_type, page, url, nil).create
   end
 
-  def self.update(params, variant_type:, page:, url:, id:)
-    new(params, variant_type, page, url, id).update
+  def self.update(params:, variant_type:, page:, id:)
+    new(params, variant_type, page, nil, id).update
   end
 
-  def initialize(params, variant_type, page, url, id = nil)
-    @params = params
+  def initialize(params, variant_type, page, url=nil, id=nil)
```

Comment:
yeah, but one of the arguments is already a parameters hash, and this just enforces that three of them are passed. I think that nesting parameter hashes deeply could get a bit hard to work with.